### PR TITLE
RUM-4116 Add the SpanLink support for Otel API implementation

### PR DESCRIPTION
**Pull request by mariusc83**
### What does this PR do?

The Otel `SpanBuilder#addLink` API delegates to the `DDSpan` under the hood and adds an `AgentSpanLink` to it. 
In this PR we are adding the links support into our `OtelDdSpanToSpanEventMapper` by serializing the `AgentSpanLink` list and adding it as a `String/String` entry into the `SpanEvent.meta` map. 
Example of a serialized span link into the `meta` section:
```
_dd.span_links -> "[
    {
        "attributes": {
            "key": "value"
        },
        "flags": 1,
        "span_id": "4ab9b106d00b9366",
        "trace_id": "000000000000000015b5cedc8e96fd32"
    }
]"

```
### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

### CODE REVIEW

File: features/dd-sdk-android-trace/src/main/kotlin/com/datadog/android/trace/internal/domain/event/OtelDdSpanToSpanEventMapper.kt [hunk 1]
```diff
@@ -104,6 +121,35 @@ internal class OtelDdSpanToSpanEventMapper(
...
+        val linkedTraceId = link.traceId().toHexString()
+        val linkedSpanId = DDSpanId.toHexStringPadded(link.spanId())
+        val attributes = toJson(link.attributes().asMap())
+        val flags = link.traceFlags()
+        val traceState = link.traceState()
+        val spanLink = JsonObject().apply {
+            addProperty(TRACE_ID_KEY, linkedTraceId)
+            addProperty(SPAN_ID_KEY, linkedSpanId)
+            add(ATTRIBUTES_KEY, attributes)
+            if (flags.toInt() == 0) {
```
Comment (mariusc83) [inline features/dd-sdk-android-trace/src/main/kotlin/com/datadog/android/trace/internal/domain/event/OtelDdSpanToSpanEventMapper.kt:143]: here it should be the opposite, I will update this now :(

**Review by xgouchet — APPROVED** [review]

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/android/trace/assertj/SpanEventAssert.kt [hunk 2]
```diff
@@ -299,6 +303,98 @@ internal class SpanEventAssert(actual: SpanEvent) :
...
+                jsonObject.addProperty(key, value)
+            }
+            val expectedAttributesJsonString = jsonObject.toString()
+            assertThat(actualAttributes)
+                .overridingErrorMessage(
+                    "Expected SpanEvent to have span link attributes: " +
+                        "$expectedAttributesJsonString but " +
+                        "instead was: $actualAttributes"
+                )
+                .isEqualTo(expectedAttributesJsonString)
```
Comment (0xnm) [inline features/dd-sdk-android-trace/src/test/kotlin/com/datadog/android/trace/assertj/SpanEventAssert.kt:393]: not blocking: maybe extract this into a dedicated assert / split assertions? a bit hard to read

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/android/utils/forge/AgentSpanLinkForgeryFactory.kt [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+            whenever(it.traceState()).doReturn(traceState)
+        }
+    }
+
+    private fun Forge.exhaustiveAttributes(): Map<String, String> {
+        return listOf(
+            aString()
+        ).map { anAlphabeticalString() to it }
+            .toMap(mutableMapOf())
+    }
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply features/dd-sdk-android-trace/src/test/kotlin/com/datadog/android/utils/forge/AgentSpanLinkForgeryFactory.kt:41]: yes that is the reason why I did not import it but you are right I need to change that name

**Review by jonathanmos — DISMISSED** [review]

File: detekt_custom.yml [hunk 4]
```diff
@@ -1120,6 +1120,7 @@ datadog:
       - "kotlin.String.substringAfter(kotlin.Char, kotlin.String)"
       - "kotlin.String.substringAfterLast(kotlin.Char, kotlin.String)"
       - "kotlin.String.substringBefore(kotlin.Char, kotlin.String)"
+      - "kotlin.String.takeLessSignificant64Bits()"
```
Comment (0xnm) [inline detekt_custom.yml:1123]:
```suggestion
      - "kotlin.String.takeLeastSignificant64Bits()"
```

**Review by 0xnm — APPROVED** [review]

**Review by 0xnm — APPROVED** [review]

**Pull request merged**